### ISSUE TITLE
Translation performance degradation after updated code

### ISSUE THREAD

**Issue by yogarshi** [OpenNMT/OpenNMT-py#329]
I was working on a Vietnamese-English translation project using the code forked on 09/14 (SHA1 - 6d56d3a401535a3cb8f1189292766ff2898909b5), and I recently decided to get more recent updates (SHA1 - 5504e10d2c1241172de71496b530aeee9fb65318). After shifting to the new version, the model seemed to do much worse during training, both on training and validation and training perplexity. The image below shows two comparative training runs - the one with the older code is on the left.

![issue](https://user-images.githubusercontent.com/3817465/31398571-654ab650-adb7-11e7-9065-cd822b62eff6.png)

Here's a snippet of my training parameters. My training and validation data is [available here](http://www.umiacs.umd.edu/~yogarshi/data/vi-en.tar.gz) for easy replication

```

python ${opennmt_dir}/preprocess.py -train_src ${data_dir}/${train_prefix}.${src} \
                                    -train_tgt ${data_dir}/${train_prefix}.${trg} \
                                    -valid_src ${data_dir}/${dev_prefix}.${src} \
                                    -valid_tgt ${data_dir}/${dev_prefix}.${trg} \
                                    -save_data ${output_dir}/${output_prefix} >> log_${seed}

python ${opennmt_dir}/train.py  -data ${output_dir}/${output_prefix} \
                                -save_model ${output_dir}/${output_prefix}_model \
                                -epochs 20 \
                                -src_word_vec_size 512 -tgt_word_vec_size 512 \
                                -global_attention general \
                                -dropout 0.4 \
                                -rnn_type GRU -rnn_size 1000 \
                                -enc_layers 1 -dec_layers 1 \ 
                                -optim adam -learning_rate 0.001 \
                                -start_decay_at 20 \
                                -batch_size 80 -report_every 1600 \
                                -gpuid 1 >> log_${seed}    
```

**Comment by JianyuZhan** [comment]
I am try reproducing, the training is really slow,  and I observed that for the old version, the valid ppl stagnates from epoch 5.  I think it is because the initial learning rate is too small(0.001),  and when trainer finds valid ppl doesn't decrease, it halves the learning rate, which pushes the learning rate to even smaller.  

Is there any justification for using a such small learning rate? If not, would you mind doing another comparison with a higher learning rate?

**Comment by yogarshi** [comment]
I think with Adam (which is what I'm using here), higher learning rates is not a good idea in general. Nevertheless,  I did a quick experiment with the old code and a higher learning rate (0.01), and the validation ppl oscillates over the first few epoch - this has the same effects as you mentioned, trainer finds valid ppl doesn't decrease and hence starts halving the learning rate.

```
Validation perplexity: 811101
Validation perplexity: 1.95345e+09
Validation perplexity: 872240
Validation perplexity: 13582.8
```

Also, I wouldn't read too much into the stagnation that you mention due to two reasons:
1) This is a reasonably small dataset by MT standards, so I'm not surprised that it converges in about 5 epochs.
2) A validation ppl of ~16 is about par for this dataset, and the final BLEU scores seem about right for the dataset.

**Comment by srush** [comment]
Sorry, your learning rate is fine. @JianyuZhan is still learning about NMT. We'll look into it more.

**Comment by JianyuZhan** [comment]
Today I parallel compared the old code and new code, especially paid attention to the `train()` and `validate()` code logic, AFAIC, no problem found.

Also, I used your data and your settings to do training, the old is the same as yours(commit 6d56d3a40153),  the new is the latest master as of Oct 11(commit 15dce1b68cde),  and interestingly I found the parallel result ppl/accuracy are quite close, and the final new ppl/accurary is even slightly better(See below).  I am doing a second run of the new code in order to exclude any randomness. It will finish by tomorrow, but looking at its current status, the result is still promising, a little bit better.

I went through all commits between your new code(commit 5504e10d2c12) and the latest master, only one PR touched this part of code, and it is also done by me, so I am quite sure it should not have modified the ppl computation logic. So I figured it might be due to some other subtle problem out of the loss computation logic that has been fixed between your new code and the latest master. And I suggest you update to latest, because it fixes several other bugs as well.


```
old(6d56d3a40153)                        new(15dce1b68cde)
================================         ================================ 
* number of parameters: 87786299	| * number of parameters: 87786299
('encoder: ', 14250544)			| ('encoder: ', 14250544)
('decoder: ', 73535755)			| ('decoder: ', 73535755)
()					|
Train perplexity: 89.2987		| Train perplexity: 84.8678
Train accuracy: 34.2669			| Train accuracy: 34.9507
Validation perplexity: 38.4203		| Validation perplexity: 31.1909
Validation accuracy: 41.7351		| Validation accuracy: 45.421
()					|
Train perplexity: 26.4087		| Train perplexity: 25.6858
Train accuracy: 46.6614			| Train accuracy: 46.9622
Validation perplexity: 22.3648		| Validation perplexity: 21.9799
Validation accuracy: 49.0842		| Validation accuracy: 49.0465
()					|
Train perplexity: 18.9884		| Train perplexity: 18.6126
Train accuracy: 50.2154			| Train accuracy: 50.4551
Validation perplexity: 19.984		| Validation perplexity: 20.2199
Validation accuracy: 50.2676		| Validation accuracy: 50.6294
()					|
Train perplexity: 15.6893		| Train perplexity: 15.4006
Train accuracy: 52.3213			| Train accuracy: 52.5251
Validation perplexity: 19.0699		| Validation perplexity: 19.9098
Validation accuracy: 51.1118		| Validation accuracy: 50.3806
()					|
Train perplexity: 13.7429		| Train perplexity: 13.5007
Train accuracy: 53.9078			| Train accuracy: 54.1101
Validation perplexity: 18.5905		| Validation perplexity: 19.16
Validation accuracy: 51.5188		| Validation accuracy: 51.1042
()					|
Train perplexity: 12.4918		| Train perplexity: 12.349
Train accuracy: 55.1554			| Train accuracy: 55.3109
Validation perplexity: 19.503		| Validation perplexity: 19.9637
Validation accuracy: 51.2399		| Validation accuracy: 50.946
Decaying learning rate to 0.0005	| Decaying learning rate to 0.0005
()					|
Train perplexity: 9.74797		| Train perplexity: 9.61541
Train accuracy: 58.8851			| Train accuracy: 59.1153
Validation perplexity: 18.0796		| Validation perplexity: 18.0501
Validation accuracy: 52.6344		| Validation accuracy: 52.966
Decaying learning rate to 0.00025	| Decaying learning rate to 0.00025
()					|
Train perplexity: 8.06284		| Train perplexity: 7.95978
Train accuracy: 61.9926			| Train accuracy: 62.2378
Validation perplexity: 18.4608		| Validation perplexity: 18.1863
Validation accuracy: 52.9208		| Validation accuracy: 53.2826
Decaying learning rate to 0.000125	| Decaying learning rate to 0.000125
()					|
Train perplexity: 7.23231		| Train perplexity: 7.12894
Train accuracy: 63.9571			| Train accuracy: 64.1914
Validation perplexity: 18.8029		| Validation perplexity: 18.6407
Validation accuracy: 52.8153		| Validation accuracy: 52.8605
Decaying learning rate to 6.25e-05	| Decaying learning rate to 6.25e-05
()					|
Train perplexity: 6.81778		| Train perplexity: 6.71182
Train accuracy: 65.0437			| Train accuracy: 65.2812
Validation perplexity: 19.0993		| Validation perplexity: 18.789
Validation accuracy: 52.6645		| Validation accuracy: 52.8002
Decaying learning rate to 3.125e-05	| Decaying learning rate to 3.125e-05
()					|
Train perplexity: 6.61177		| Train perplexity: 6.50805
Train accuracy: 65.6083			| Train accuracy: 65.8697
Validation perplexity: 19.2717		| Validation perplexity: 19.0092
Validation accuracy: 52.6494		| Validation accuracy: 52.8529
Decaying learning rate to 1.5625e-05	| Decaying learning rate to 1.5625e-05
()					|
Train perplexity: 6.50125		| Train perplexity: 6.41118
Train accuracy: 65.9415			| Train accuracy: 66.1759
Validation perplexity: 19.3744		| Validation perplexity: 19.0178
Validation accuracy: 52.6268		| Validation accuracy: 52.9132
Decaying learning rate to 7.8125e-06	| Decaying learning rate to 7.8125e-06
()					|
Train perplexity: 6.45362		| Train perplexity: 6.36154
Train accuracy: 66.0896			| Train accuracy: 66.3146
Validation perplexity: 19.3997		| Validation perplexity: 19.0625
Validation accuracy: 52.672		| Validation accuracy: 52.9434
Decaying learning rate to 3.90625e-06	| Decaying learning rate to 3.90625e-06
()					|
Train perplexity: 6.42861		| Train perplexity: 6.33128
Train accuracy: 66.1503			| Train accuracy: 66.4002
Validation perplexity: 19.4222		| Validation perplexity: 19.1222
Validation accuracy: 52.6193		| Validation accuracy: 52.9132
Decaying learning rate to 1.95313e-06	| Decaying learning rate to 1.95313e-06
()					|
Train perplexity: 6.41743		| Train perplexity: 6.32036
Train accuracy: 66.2027			| Train accuracy: 66.4534
Validation perplexity: 19.4111		| Validation perplexity: 19.13
Validation accuracy: 52.657		| Validation accuracy: 52.9585
Decaying learning rate to 9.76563e-07	| Decaying learning rate to 9.76563e-07
()					|
Train perplexity: 6.40613		| Train perplexity: 6.31449
Train accuracy: 66.2376			| Train accuracy: 66.4581
Validation perplexity: 19.4399		| Validation perplexity: 19.1447
Validation accuracy: 52.6419		| Validation accuracy: 52.9509
Decaying learning rate to 4.88281e-07	| Decaying learning rate to 4.88281e-07
()					|
Train perplexity: 6.4059		| Train perplexity: 6.31177
Train accuracy: 66.226			| Train accuracy: 66.4678
Validation perplexity: 19.4495		| Validation perplexity: 19.152
Validation accuracy: 52.6268		| Validation accuracy: 52.9283
Decaying learning rate to 2.44141e-07	| Decaying learning rate to 2.44141e-07
()					|
Train perplexity: 6.40267		| Train perplexity: 6.31153
Train accuracy: 66.2199			| Train accuracy: 66.4807
Validation perplexity: 19.4525		| Validation perplexity: 19.1494
Validation accuracy: 52.6117		| Validation accuracy: 52.9359
Decaying learning rate to 1.2207e-07	| Decaying learning rate to 1.2207e-07
()					|
Train perplexity: 6.40756		| Train perplexity: 6.31039
Train accuracy: 66.1984			| Train accuracy: 66.4879
Validation perplexity: 19.4544		| Validation perplexity: 19.1496
Validation accuracy: 52.6117		| Validation accuracy: 52.9434
Decaying learning rate to 6.10352e-08	  Decaying learning rate to 6.10352e-08
()					|
Train perplexity: 6.40812		| Train perplexity: 6.30794
Train accuracy: 66.2099			| Train accuracy: 66.5031
Validation perplexity: 19.455		| Validation perplexity: 19.1496
Validation accuracy: 52.6268		| Validation accuracy: 52.9434
Decaying learning rate to 3.05176e-08	| Decaying learning rate to 3.05176e-08
```

**Comment by yogarshi** [comment]
Seems that pulling to the latest master has made the problem go away. I'm going to close the issue for now. Thanks for the help @JianyuZhan and @srush!

**Comment by srush** [comment]
Thanks so much @yogarshi , don't hesitate to file more issues as they come up.
